engine/table/src/main/java/io/deephaven/engine/table/impl/replay/ReplayTable.java

```diff
@@ -4,69 +4,83 @@
-import io.deephaven.engine.rowset.RowSetBuilderRandom;
-import io.deephaven.engine.rowset.RowSetFactory;
+import io.deephaven.util.QueryConstants;
+import org.jetbrains.annotations.NotNull;
 
 import java.util.Map;
 
 public class ReplayTable extends QueryTable implements Runnable {
 
+    private final Replayer replayer;
+    private final ColumnSource<Long> nanoTimeSource;
+    private final RowSet.Iterator rowSetIterator;
 
-    private final RowSet.Iterator indexIterator;
-    private long curr;
-    private final ColumnSource<DateTime> timeSource;
+    private long nextRowKey = RowSequence.NULL_ROW_KEY;
+    private long nextTimeNanos = QueryConstants.NULL_LONG;
     private boolean done;
-    private final Replayer replayer;
 
-    public ReplayTable(RowSet rowSet, Map<String, ? extends ColumnSource<?>> result, String timeColumn,
-            Replayer replayer) {
-        super(RowSetFactory.empty().toTracking(), result);
-        Require.requirement(replayer != null, "replayer != null");
-        // noinspection unchecked
-        replayer.registerTimeSource(rowSet, (ColumnSource<DateTime>) result.get(timeColumn));
+    public ReplayTable(
+            @NotNull final RowSet rowSet,
+            @NotNull final Map<String, ? extends ColumnSource<?>> columns,
+            @NotNull final String timeColumn,
+            @NotNull final Replayer replayer) {
+        super(RowSetFactory.empty().toTracking(), columns);
+        this.replayer = Require.neqNull(replayer, "replayer");
+        // NB: This will behave incorrectly if our row set or any data in columns can change. Our source table *must*
+        // be static. We also seem to be assuming that timeSource has no null values in rowSet. It would be nice to use
+        // a column iterator for this, but that would upset unit tests by keeping pooled chunks across cycles.
+        final ColumnSource<DateTime> dateTimeSource = getColumnSource(timeColumn, DateTime.class);
+        replayer.registerTimeSource(rowSet, dateTimeSource);
+        nanoTimeSource = ReinterpretUtils.dateTimeToLongSource(dateTimeSource);
+        rowSetIterator = rowSet.iterator();
+
         setRefreshing(true);
-        indexIterator = rowSet.iterator();
-        if (indexIterator.hasNext()) {
-            curr = indexIterator.nextLong();
+
+        advanceIterators();
+        if (!done) {
+            try (final RowSet initial = advanceToCurrentTime()) {
+                getRowSet().writableCast().insert(initial);
+            }
+        }
+    }
+
+    private void advanceIterators() {
+        if (rowSetIterator.hasNext()) {
+            nextRowKey = rowSetIterator.nextLong();
+            nextTimeNanos = nanoTimeSource.getLong(nextRowKey);
         } else {
+            // NB: It would be best to ensure that if this is hit during construction, we're never added to the UGP.
+            // If this is hit during update processing, it would be great to remove ourselves from the UGP right away.
```

Comment:
1) Initials in comments
2) Should this change be made?  @rcaudy

Reply:
This change can't be made without a refactoring of the Replayer.